### Romberg method

See in *PDF* folder for precise symbolic expressions

In [1]:
import numpy as np
from scipy.integrate import quad

In [2]:
def integral(f, a, b, h): # Order 2 integration method

    ba = b - a
    N = round(ba / h)
    h = ba / N # Recalculate h to match the exact grid
    x = np.linspace(a, b, N, endpoint = False)
    midpoints = x + h / 2.0

    I = h * np.sum(f(midpoints))

    return I

In [3]:
f = lambda x: np.sqrt(x * np.tan(x)) # f(x) = sqrt(x * tan(x))
a, b = 0.0, np.pi / 3.0 # a = 0, b = pi / 3

h = 1e-02 # h = 0.01, not so small
N = 6
Ih = []
for _ in range(N + 1): # Computing I(h), I(h/2), I(h/4), ...
    x = np.arange(a, b + h, h)
    Ih.append(
        integral(
            f = f,
            a = a,
            b = b,
            h = h
        )
    )

    h /= 2.0

Ih

[np.float64(0.6148719351566976),
 np.float64(0.6148756472806158),
 np.float64(0.6148765887600253),
 np.float64(0.6148768226390339),
 np.float64(0.6148768811089709),
 np.float64(0.6148768957235581),
 np.float64(0.6148768993801141)]

Do just 5 extrapolations. Take $I_{romb}^{4}$ as the best result. The final absolute error can be estimated as $ \left| I_{romb}^{5} - I_{romb}^{4} \right|$

In [4]:
I_romb1 = (- Ih[0] + 4.0 * Ih[1]) / 3.0 # One extrapolation
I_romb1

np.float64(0.6148768846552553)

In [5]:
I_romb2 = (Ih[0] - 20.0 * Ih[1] + 64.0 * Ih[2]) / 45.0 # Two extrapolations
I_romb2, abs(I_romb2 - I_romb1) # Absolute error of I_romb1

(np.float64(0.6148769037819111), np.float64(1.9126655814716287e-08))

In [6]:
I_romb3 = (- Ih[0] + 84.0 * Ih[1] - 1344.0 * Ih[2] + 4096.0 * Ih[3]) / 2835.0 # Three extrapolations
I_romb3, abs(I_romb3 - I_romb2) # Absolute error of I_romb2

(np.float64(0.6148769004135534), np.float64(3.3683577038701173e-09))

In [7]:
I_romb4 = (Ih[0] - 340.0 * Ih[1] + 22848.0 * Ih[2] - 348160.0 * Ih[3] + 1048576.0 * Ih[4]) / 722925.0 # Four extrapolations
I_romb4, abs(I_romb4 - I_romb3) # I_romb4 is our best result. # abs(I_romb4 - I_romb3) is the absolute error of I_romb3

(np.float64(0.6148769006018093), np.float64(1.8825596637128683e-10))

In [8]:
I_romb5 = (- Ih[0] + 1364.0 * Ih[1] - 371008.0 * Ih[2] + 23744512.0 * Ih[3] - 357564416.0 * Ih[4] + 1073741824.0 * Ih[5]) / 739552275.0 # Five extrapolations
I_romb5, abs(I_romb5 - I_romb4) # Absolute error of I_romb4

(np.float64(0.6148769005947323), np.float64(7.077005648170598e-12))

In [9]:
real, err = quad(f, a, b) # Precise value
real, err

(0.6148769005989663, 6.826504923551261e-15)

In [10]:
# Romb4 is much better than I_h
abs(I_romb4 - real), abs(Ih[0] - real)

(np.float64(2.843059121460101e-12), np.float64(4.965442268689557e-06))

Best Romberg result computed:

$$
    \int_0^{\frac{\pi}{3}} \, \sqrt{x \, \tan{x}} \, dx = 0.614876900602 \pm 0.000000000007
$$

Let's reduce $h$

In [11]:
h = 1e-03 # h = 0.001
Ih = []
for _ in range(N + 1): # Computing I(h), I(h/2), I(h/4), ...
    x = np.arange(a, b + h, h)
    Ih.append(
        integral(
            f = f,
            a = a,
            b = b,
            h = h
        )
    )

    h /= 2.0

Ih

[np.float64(0.6148768506569285),
 np.float64(0.6148768881134519),
 np.float64(0.6148768974790774),
 np.float64(0.614876899818994),
 np.float64(0.6148769004039499),
 np.float64(0.6148769005502123),
 np.float64(0.614876900586778)]

In [12]:
I_romb4 = (Ih[0] - 340.0 * Ih[1] + 22848.0 * Ih[2] - 348160.0 * Ih[3] + 1048576.0 * Ih[4]) / 722925.0 # Four extrapolations
I_romb5 = (- Ih[0] + 1364.0 * Ih[1] - 371008.0 * Ih[2] + 23744512.0 * Ih[3] - 357564416.0 * Ih[4] + 1073741824.0 * Ih[5]) / 739552275.0 # Five extrapolations

I_romb4, abs(I_romb5 - I_romb4) # Result, error

(np.float64(0.6148769005989353), np.float64(3.3861802251067274e-14))

Result:

$$
    \int_0^{\frac{\pi}{3}} \, \sqrt{x \, \tan{x}} \, dx = 0.61487690059894 \pm 0.00000000000003 
$$

In [13]:
# Real error committed
abs(I_romb4 - real)

np.float64(3.097522238704187e-14)

Let's reduce $h$ again. See what happens

In [14]:
h = 25e-07 # h = 0.0000025
Ih = []
for _ in range(N + 1): # Computing I(h), I(h/2), I(h/4), ...
    x = np.arange(a, b + h, h)
    Ih.append(
        integral(
            f = f,
            a = a,
            b = b,
            h = h
        )
    )

    h /= 2.0

I_romb4 = (Ih[0] - 340.0 * Ih[1] + 22848.0 * Ih[2] - 348160.0 * Ih[3] + 1048576.0 * Ih[4]) / 722925.0 # Four extrapolations
I_romb5 = (- Ih[0] + 1364.0 * Ih[1] - 371008.0 * Ih[2] + 23744512.0 * Ih[3] - 357564416.0 * Ih[4] + 1073741824.0 * Ih[5]) / 739552275.0 # Five extrapolations

I_romb4, abs(I_romb5 - I_romb4) # Result, error

(np.float64(0.6148769005989663), np.float64(1.1102230246251565e-16))

Result:

$$
    \int_0^{\frac{\pi}{3}} \, \sqrt{x \, \tan{x}} \, dx = 0.6148769005989663 \pm 0.0000000000000001
$$

We reached the maximum accuracy possible. Machine precision

In [15]:
np.finfo(np.float64).eps / 2.0 # Machine precision

np.float64(1.1102230246251565e-16)

In [16]:
abs(I_romb5 - I_romb4) / I_romb4 # Relative error near machine precision

np.float64(1.8056021027032595e-16)